protocol/src/main/java/org/geysermc/mcprotocollib/protocol/ServerListener.java

```diff
@@ -125,8 +125,7 @@ public void packetReceived(Session session, Packet packet) {
                 PrivateKey privateKey = KEY_PAIR.getPrivate();
 
                 if (!Arrays.equals(this.challenge, keyPacket.getEncryptedChallenge(privateKey))) {
-                    session.disconnect("Invalid challenge!");
-                    return;
+                    throw new IllegalStateException("Protocol error");
```

Comment:
Why are we not disconnecting the session here first?

Reply:
It's what vanilla does. It's seen as an unexpected error because the client sent an invalid payload.

Reply:
Will the thrown exception result in a session disconnect? If not, i'd definitely do that first before throwing/returning
IMO we don't need to follow vanilla exactly, especially for cases like these

Reply:
Depends whether it runs on the netty thread or the event loop. Worst case, it just times out and prints a error in the console.

Reply:
It's basically the same level as us throwing decryption errors, which the key packet already does.

Reply:
A timeout seems sub-optimal; i'd rather keep the current disconnect behavior to ensure that cannot happen

Reply:
Like for example `getEncryptedChallenge` could error with an `IllegalStateException` (Explicitly thrown too) because the bytes are incorrect. Same goes for this case.

Reply:
Actually, NVM. It doesn't depend on the loop. It'll cause a immediate disconnect with an "Internal Exception" message with an error message. 
![image](https://github.com/GeyserMC/MCProtocolLib/assets/40795980/ed3fd809-4ad1-4493-92e1-d6d39bc442f8)

Reply:
Okay, that sounds fair too; but let's hear what others think

Reply:
Seems alright - looks like any other SessionListeners will be skipped which is good. Keeping the current message seems more developer friendly though

Reply:
yup, I don't think branching off vanilla behavior benefits anyone here. It makes comparing the code directly with vanilla harder for an edge case that'll basically be never be called.